Update enable forging guide (#1654)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/understand-blockchain/index.adoc

== What are sidechain applications?

A *sidechain application* (or, in short: *sApp*) refers to a software that contains a lisk based <<sidechain-client,sidechain client>>, a <<middleware,middleware>> and a <<ui,user interface>>.

Unlike smart contract based platforms, where smart contracts have to share the same chain for their blockchain needs, Lisk provides the opportunity for each application to run on its own blockchain.

This provides faster block generation, reduces the burden on the network, and increases scalability as discussed on the xref:{url_scalability_introduction}[Blockchain Scalability] page.
Sidechains can communicate with other sidechains with the help of xref:{lisk_cross_chain_communication}[Lisk Cross Chain Communication].

[#sidechain-client]
Sidechain client::
The sidechain client is the central piece of software that is needed to establish and maintain a blockchain network between many different independent servers.

> Servers with running sidechain clients are called *sidechain nodes* in this context.

Author: ```suggestion
Servers with running sidechain clients are called *sidechain nodes*.
```

//TODO: Update diagram to be fully visible with dark theme
.Architecture of a Lisk sidechain client
image::architecture.png[lisk-framework-architecture, 700 , align="center"]

=== Application layer
The application layer handles state changes to the sidechain.
The function of the application layer is to act as an interface to connect to the outside world, such as various external services in order to send and receive data.
An application layer consists of a <<State machine>>, <<Modules>>, <<Plugins>>, and <<Configuration>>.

==== State machine

As the name suggests, a xref:{url_understand_state_machine}[state machine] is relevant to the states of a machine.
A sidechan client relies heavily on its state machine to mutate the state of a sidechain.

* *States:* A state machine is deterministic and can have multiple states, but only one state at any given time.
In the context of the Lisk blockchain, a *key-value-store* represents the current state of the blockchain, containing all on-chain data of the blockchain.

* *Transitions*: A transition is defined as the instantaneous transfer from one state to another state.

> In the context of the Lisk blockchain, a *transition* of states is triggered through *blocks* and their containing *transactions* which are added to the blockchain.

Author: ```suggestion
In the context of the Lisk blockchain, a *transition* of the state is triggered through *blocks* and the *transactions* present in those blocks; i.e. every new block that is added to the blockchain mutates the state of the blockchain.
```

==== Modules

xref:{url_introduction_modules}[Modules] aid the state machine to transition the state of the sidechain with verified and validated data.
They contain *on-chain logic* which is part of the sidechain protocol.

For example, if Bob wants to send 10 LSK tokens to Alice then, behind the scenes a module will verify the validity of such a request.
Upon validation and verification, the module will ask the state machine to transfer 10 LSK tokens from Bob's account to Alice's account.

// TIP: Lisk provides a range of default modules out of the box. These modules are created automatically whenever a sidechain application is bootstrapped via Lisk Commander. For more information, see the xref:{url_lisk_default_modules}[].
TIP: Lisk provides a range of default modules out of the box.
These modules are used automatically, whenever a sidechain application is xref:{url_build_index}[bootstrapped via Lisk Commander].

===== When to create a module

Modules are able to perform the following functions:

* Define how data is stored on the sidechain.

> * Define logic that is executed per block footnote:block_footnote[For more information about blocks, check out xref:{url_protocol_blocks}[Blocks].]

Reviewer: ```suggestion
* Define logic that is executed per block footnote:block_footnote[For more information about blocks, check out xref:{url_protocol_blocks}[Blocks].].
```

// A chain is also responsible for fetching blockchain data from the data stores.

==== P2P
The P2P component handles sending and receiving data from nodes. 
It also maintains an active connection with the Lisk network. 
Every node receives new blocks generated by other nodes via the P2P network.
The receiver node in that case repeats all the steps mentioned in the <<Engine layer>>. 
If the received block is verified, then the receiving node adds it to its blockchain instance.

==== Data Stores

> Each blockchain node is an instance of a particular sidechain and keeps the data on various data stores inside.

Reviewer: ```suggestion
Each blockchain node is an instance of a particular sidechain and each node keeps its data on various data stores.
```

> This data contains the following two types: on-chain and off-chain.

Reviewer: ```suggestion
This data is of the following two types: on-chain and off-chain.
```

Lisk also maintains off-chain data regarding the legacy chains.
Legacy data consists of blocks from depreciated versions of the protocol.

=== Communication interfaces / APIs

.RPC Communication of a Lisk node
image::intro/communication-architecture.png[communication-interfaces,1000 ,align="center"]

The communication architecture of a sidechain client allows internal application components and external services to communicate to the client via various channels.

Sidechain clients support three industry-standard communication protocols:

. Inter-Process Communication (IPC)
. Web Sockets (WS), and
. Hypertext Transfer Protocol (HTTP).

> The communication protocol of a sidechain client can be changed through xref:{url_advanced_architecture_config}[configurations].

Reviewer: The URL takes you to v5 docs, I think it should go to v6 docs.

The frontend can then request this data via API requests from Lisk Service.

// TODO: this will be revised after terminology discussion.
////

> == Sidechain applications vs. dApps

Reviewer: Why have we commented this section?

Author: I commented it, because we didn't update this section yet, it is also indicated by the TODO above.
Once the content is updated we can add it back in :)
